### ISSUE TITLE
Nullability friendly builder

### ISSUE THREAD

**Issue by bangarharshit** [Microsoft/thrifty#162]
Most of the static code analysis tools such as NullAway and infer throws error with the current builder setup.

There are 2 issues:
1. Builder fields are not annotated.
2. Builder fields can't be non-null with the current setup since the constructor is usually empty.

```java
public Builder() {}

public Builder nonNullField(@NonNull String nonNullField) {}

public Builder nullableField(@Nullable String nullableField) {}
```
 An alternate way will be to set all the required values in Builder's constructor and only have separate methods for Nullable fields.

```java
public Builder(@NonNull String nonNullField) {}

public Builder nullableField(@Nullable String nullableField) {}
```

Since this breaks backward compatibility it needed to be flagged.

Complete class - https://gist.github.com/bangarharshit/c4ec0bee365cc4ef30270f10d4d1aa2f

**Comment by benjamin-bader** [comment]
I think annotating parameters of setter methods is a great idea - I'm surprised we don't already do that!  I'm not familiar with errors from infer or nullaway - which parts of the generated code get flagged?

I'd rather not force builders to be "correct" on construction; they are intended to be a readable way to build up their companion struct, and stuffing all the required params in their constructor would mostly defeat that purpose.

**Comment by bangarharshit** [comment]
That is a very good point. Here is a demo [project](https://github.com/bangarharshit/Nullness-Demo/tree/thrity) with readme containing all the error details. 

An alternative will be to use some kind of `required` annotation on top of builder methods. Let me surface that to these checker teams.

Adding annotation to both parameters and fields will still be a very useful.

**Comment by benjamin-bader** [comment]
Thanks for the demo project!  I see that NullAway indeed doesn't like the builder fields being without annotations.

I think it's interesting that there's no warning in the `Arbit` constructor - looks like the tool is smart enough to see that the ctor is manually checking for null and throwing?  In that case, how about just annotating every builder field as `@Nullable`, and annotating the setter parameters according to the field's requiredness?  e.g.

```java
class ArbitBuilder {
  @Nullable String requiredField = null;
  @Nullable String optionalField = null;

  ArbitBuilder setRequiredField(@NonNull String required) {
    ...
  }

  ArbitBuilder setOptionalField(@Nullable String optionalField) {
    ...
  }
}
```

This way, the tool would be satisfied, and the annotations themselves would still be honest.  Does that sound like it would work?  What do you think?

**Comment by bangarharshit** [comment]
Verified and you are right. Just annotating the fields in the way you suggested works.  Can send a PR with relevant annotations.

**Comment by benjamin-bader** [comment]
Fixed in #164.
